README.md
# polyserve

[![Build Status](https://travis-ci.org/PolymerLabs/polyserve.svg?branch=master)](https://travis-ci.org/PolymerLabs/polyserve)

A simple development server for web projects.

`polyserve` serves project files from a URL root that allows relative URLs
that reach out of the project, like those starting with `../`, to work. This is
necessary for referencing other packages by path when stored as a flat directory
such as how Bower works.

The local package is served at the URL `/components/{bower-name}/`, with files
served from the current directory. Other packages are served at
`/components/{packageName}` with files served from their directory under
`./bower_components/`.

## Installation

    $ npm install polyserve -g

## Usage

### Run `polyserve`

    $ cd my-element/
    $ polyserve

### Browse files

Navigate to `localhost:8080/components/my-element/demo.html`

### Options

  * `-c` <component-dir> The component directory to use. Defaults to reading from the Bower config (usually bower_components/)
  * `-n` Package name. Defaults to reading from bower.json
  * `-p` The TCP port to use for the web server
  * `-o` Opens your default browser to an initial page, e.g. "demo" or "index.html"
  * `-b <browsername>` use this browser instead of default (ex: 'Google Chrome Canary')
  * `-H <hostname>` use this hostname instead of 0.0.0.0 (all IP addresses) 
  * `-P <protocol>` The server protocol to use {`h2`, `https/1.1`, `http/1.1`}. **`h2` requires Node 5+.**
  * `-key <path>` Path to TLS certificate private key file for https. Defaults to "key.pem".
  * `-cert <path>` Path to TLS certificate file for https. Defaults to "cert.pem".
  * `-manifest <path>` Path to h2-push manifest

## Compiling from Source

    $ npm install
    $ npm run build

You can compile and run polyserve from source by cloning the repo from Github and then running `npm run build`. Make sure you have already run `npm install` before building.

### Run Tests

    $ npm test


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

<!-- ## Unreleased -->

* ES > AMD module transformation
  * AMD loader will now be injected as a minified inline script, instead of as
    an external script.
  * AMD modules will now execute in document script order.
* Babel helpers will now be injected as a single minified inline script into
  the HTML document, instead of into every JS script.
* Phantom `<html>`, `<body>`, and `<head>` elements will no longer be
  introduced into HTML.

## [0.25.3](https://github.com/PolymerLabs/polyserve/tree/0.25.3) (2018-03-26)

* Node module specifier rewriter will now resolve paths according to the node
  module resolution algorithm, not just bare specifiers (e.g. "./foo" resolves
  to "./foo.js").

## [0.25.2](https://github.com/PolymerLabs/polyserve/tree/0.25.1) (2018-03-21)

* Another fix for import specifier rewriting when given a relative componentDir.

## [0.25.1](https://github.com/PolymerLabs/polyserve/tree/0.25.1) (2018-03-21)

* Fix import specifier rewriting when given a relative componentDir.

## [0.25.0](https://github.com/PolymerLabs/polyserve/tree/0.25.0) (2018-03-20)

* Module bare specifier rewriting Babel plugin now works on Windows, does not rewrite fully qualified URLs, and will follow the "module" or "jsnext:main" fields when a package.json uses them instead of "main".
* Add exponentiation, async/await, and async generator support when compiling to ES5.
* Fix import specifier rewriting when importing a dependency from a top-level module.

## [0.24.0](https://github.com/PolymerLabs/polyserve/tree/0.24.0) (2018-02-23)

* Honor `.bowerrc` for bower component directory and variants
* Root directories can now be relative paths
* Added --module-resolution flag to allow rewriting of "bare" module specifiers to web-compatible paths.

## [0.23.0](https://github.com/PolymerLabs/polyserve/tree/0.23.0) (2017-10-02)

* Added ability to modify and/or substitute generated express apps for when calling startServers, to support more flexibility in request handler ordering and middleware scenarios.
* Fix issue with module scripts being compiled when a nomodule script exists.
* Updated polymer-build to 2.1.0 which includes polymer-build 3.1.0 which inlines external stylesheet links in templates.

## [0.22.1](https://github.com/PolymerLabs/polyserve/tree/v0.22.1) (2017-09-19)

* Fixes issue with the babel compile cache where different compilation options were using the same namespace. ([#216](https://github.com/Polymer/polyserve/issues/216)).

## [0.22.0](https://github.com/PolymerLabs/polyserve/tree/v0.22.0) (2017-09-18)

* Identify wct-browser-legacy as a web-component-tester client package, to add a hook for deferring mocha execution in support of requirejs.
* Change `x(...args)` syntax to `x.apply(undefined, arguments)` to support IE11.

## [0.21.9](https://github.com/PolymerLabs/polyserve/tree/v0.21.9) (2017-09-15)

* Fix issue where requirejs is installed somewhere other than in polyserve's own node_modules subfolder.
* Fix issue with `.npmignore` file and npm pack logic change in recent npm release.

## [0.21.0](https://github.com/PolymerLabs/polyserve/tree/v0.21.0) (2017-09-13)

* Auto-compile now includes transformation of ES modules to AMD modules with RequireJS for browsers that do not support ES modules. Includes special handling for WCT.
* Fix #48 where we would incorrectly inject the ES5 adapter for browsers that supported ES2015 but not modules.

## [0.20.0](https://github.com/PolymerLabs/polyserve/tree/v0.20.0) (2017-08-08)

* Update Polymer Build to v2.0.0
* Allow package names to include a '/'
* Added --npm flag which sets component folder to node_modules and reads package name from package.json.

## [0.19.1](https://github.com/PolymerLabs/polyserve/tree/v0.19.1) (2017-05-22)

* Allow webcomponentsjs polyfill *tests* to be compiled.

## [0.19.0](https://github.com/PolymerLabs/polyserve/tree/v0.19.0) (2017-05-08)

* Add auto-compile support for the Mobile Safari browser.
* Add auto-compile support for the Vivaldi browser.
* Fixed issue when serve with --push-manifest ([#168](https://github.com/Polymer/polyserve/issues/168))
* Add gzip and deflate HTTP response compression.

## [0.18.0](https://github.com/PolymerLabs/polyserve/tree/v0.18.0) (2017-04-18)

* When compiling to ES5, inject the Custom Elements ES5 Adapter into any HTML
  file where the web components polyfill is included (typically the entry
  point). This adapter is needed when serving ES5 to browsers that support the
  native Custom Elements API
  ([#164](https://github.com/Polymer/polyserve/issues/164)).
* A server can now be started with a custom entry point; previously index.html
  was always assumed ([#161](https://github.com/Polymer/polyserve/issues/161)).

## [0.17.0](https://github.com/PolymerLabs/polyserve/tree/v0.17.0) (2017-04-13)

* Add auto-compile support for the Chromium browser.
* Add auto-compile support for the Opera browser.
* Require ES5 compilation for Edge < 40 ([#161](https://github.com/Polymer/polyserve/issues/161)).
* Fixed issue where directory paths redirected to the file system ([#96](https://github.com/Polymer/polyserve/issues/96)).
* [Breaking] Remove Node v4 support: Node v4 is no longer in Active LTS, so as per the [Polymer Tools Node.js Support Policy](https://www.polymer-project.org/2.0/docs/tools/node-support) polyserve will not support Node v4. Please update to Node v6 or later to continue using the latest verisons of Polymer tooling.

## [0.16.0](https://github.com/PolymerLabs/polyserve/tree/v0.16.0) (2017-02-14)

### Added
* `-u`/`--component-url` option to support expressing different url to
  fetch components from than `components`.

### Fixed
* When no port is given, do a better job of finding an available port.

src/args.ts
/**
 * @license
 * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
 * This code may only be used under the BSD style license found at
 * http://polymer.github.io/LICENSE.txt
 * The complete set of authors may be found at
 * http://polymer.github.io/AUTHORS.txt
 * The complete set of contributors may be found at
 * http://polymer.github.io/CONTRIBUTORS.txt
 * Code distributed by Google as part of the polymer project is also
 * subject to an additional IP rights grant found at
 * http://polymer.github.io/PATENTS.txt
 */

/// <reference path="../custom_typings/command-line-args.d.ts" />
import {ArgDescriptor} from 'command-line-args';

export let args: ArgDescriptor[] = [
  {
    name: 'version',
    description: 'Print version info',
    type: Boolean,
  },
  {
    name: 'root',
    description: 'The root directory of your project. Defaults to the current' +
        ' working directory',
    type: String,
    defaultOption: true,
  },
  {
    name: 'compile',
    description: 'Compiler options. Valid values are "auto", "always" and ' +
        '"never". "auto" compiles JavaScript to ES5 for browsers that don\'t ' +
        'fully support ES6.',
    type: String,
    defaultValue: 'auto',
  },
  {
    name: 'compile-cache',
    description: 'Maximum size in bytes (actually, UTF-8 characters) of ' +
        'the cache used to store results for JavaScript compilation. Cache ' +
        'size includes the uncompiled and compiled file content lengths. ' +
        'Defaults to 52428800 (50MB)',
    type: Number,
    defaultValue: 52428800,
  },
  {
    name: 'port',
    alias: 'p',
    description: 'The port to serve from. Defaults to 8080',
    type: Number,
  },
  {
    name: 'hostname',
    alias: 'H',
    description: 'The hostname to serve from. Defaults to localhost',
    type: String,
  },
  {
    name: 'component-dir',
    alias: 'c',
    description: 'The component directory to use. Defaults to reading from' +
        ' the Bower config (usually bower_components/)',
    type: String,
  },
  {
    name: 'component-url',
    alias: 'u',
    description: 'The component url to use. Defaults to reading from' +
        ' the Bower config (usually bower_components/)',
    type: String,
  },
  {
    name: 'package-name',
    alias: 'n',
    description: 'The package name to use for the root directory. Defaults to' +
        ' reading from bower.json',
    type: String,
  },
  {
    name: 'open',
    alias: 'o',
    description: 'The page to open in the default browser on startup.',
    type: Boolean,
  },
  {
    name: 'browser',
    alias: 'b',
    description: 'The browser(s) to open with when using the --open option.' +
        ' Defaults to your default web browser.',
    type: String,
    multiple: true,
  },
  {
    name: 'open-path',
    description: 'The URL path to open when using the --open option.' +
        ' Defaults to "index.html".',
    type: String,
  },
  {
    name: 'protocol',
    alias: 'P',
    description: 'The server protocol to use {h2, https/1.1, http/1.1}.' +
        ' Defaults to "http/1.1".',
    defaultValue: 'http/1.1',
    type: String,
  },
  {
    name: 'key',
    description: 'Path to TLS certificate private key file for https.' +
        ' Defaults to "key.pem".',
    defaultValue: 'key.pem',
    type: String,
  },
  {
    name: 'cert',
    description: 'Path to TLS certificate file for https.' +
        ' Defaults to "cert.pem".',
    defaultValue: 'cert.pem',
    type: String,
  },
  {
    name: 'manifest',
    description: 'Path to h2-push manifest',
    type: String,
  },
  {
    name: 'proxy-path',
    description:
        'Top-level path that should be redirected to the proxy-target. ' +
        'E.g. `api/v1` when you want to redirect all requests of ' +
        '`https://localhost/api/v1/`',
    type: String,
  },
  {
    name: 'proxy-target',
    description:
        'Host URL to proxy to, for example `https://myredirect:8080/foo`',
    type: String,
  },
  {
    name: 'live-reload-path',
    description: 'The file path to watch for changes to automatically reload connected browsers.',
    type: String,
  },
];
